### ISSUE TITLE
Using Azure Active Directory OAuth with Identity ASP.NET Core 2.0

### ISSUE THREAD

**Issue by AfkAtm** [aspnet/Identity#1498]
As part of the built in libraries that comes with ASP.NET Core 2.0 there is the ability to add OAuth using Microsoft accounts into the identity model.  The same way you can add Facebook, Google or Twitter.  I've used these before without issues and they are really easy to set up.  However, I've ran into no end of trouble trying to implement Azure Active Directory OAuth tying into the Identity model.  

I'm not sure what I'm doing wrong as I've modeled my custom classes off of other OAuth providers and have done a great deal of research into AAD OAuth and interacting with it.  However, no matter what I do my ClaimsIdentity ends up being empty and SigninManager returns null when trying to pick up ExternalLogin information.  

I've given more details on my issue over on Stack Overflow as I'm not sure if there is some misunderstanding of how this code should work on my end or if there is an issue with identity and playing with Azure Active Directory (which is why I'm adding this issue).  
https://stackoverflow.com/questions/47188066/using-azure-active-directory-oauth-with-identity-model-in-asp-net-core-2-0

**Comment by HaoK** [comment]
What's your startup like?

**Comment by AfkAtm** [comment]
Since I had to write some custom code to handle the request to the auth and token endpoints there is quite a lot of code related to this issue which is all included on the Stack Overflow question.  Here's the start-up piece that adds the OAuth handler to the AuthenticationBuilder:
`services.AddAuthentication()
.AddAzureAd(options =>
{
    options.ClientId = Configuration["AzureAd:ClientId"]; 
    options.AuthorizationEndpoint = $"{Configuration["AzureAd:Instance"]}{Configuration["AzureAd:TenantId"]}/oauth2/authorize"; 
    options.TokenEndpoint = $"{Configuration["AzureAd:Instance"]}{Configuration["AzureAd:TenantId"]}/oauth2/token"; 
    options.UserInformationEndpoint = $"{Configuration["AzureAd:Instance"]}{Configuration["AzureAd:TenantId"]}/openid/userinfo"; 
    options.Resource = Configuration["AzureAd:ClientId"]; 
    options.ClientSecret = Configuration["AzureAd:ClientSecret"]; 
    options.CallbackPath = Configuration["AzureAd:CallbackPath"]; 
});`

**Comment by HaoK** [comment]
Yeah that doesn't really matter to identity, are you just using the normal AddIdentity(), are you seeing the normal external redirection and bounce back to your external login callback?  Are you seeing an external cookie sent?  That's what the GetExternalLoginInfo call is basically lookin for.

**Comment by AfkAtm** [comment]
Here's the Identity and Cookie part of the Start-Up, it is pretty much just the normal AddIdentity with custom objects mixed in.  

```
            services.AddIdentity<ApplicationUser, IdentityRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>()
                .AddDefaultTokenProviders();

            services.ConfigureApplicationCookie(options =>
            {
                // Cookie settings
                options.Cookie.HttpOnly = true;
                options.Cookie.Expiration = TimeSpan.FromDays(150);
                options.LoginPath = "/Account/SignIn"; // If the LoginPath is not set here, ASP.NET Core will default to /Account/Login
                options.LogoutPath = "/Account/SignOut"; // If the LogoutPath is not set here, ASP.NET Core will default to /Account/Logout
                options.AccessDeniedPath = "/Account/AccessDenied"; // If the AccessDeniedPath is not set here, ASP.NET Core will default to /Account/AccessDenied
                options.SlidingExpiration = true;
            });
```

>Are you seeing the normal external redirection and bounce back to your external login callback?

Yes
>Are you seeing an external cookie sent?

Yes

Here are screenshots of the flow:
>SignIn

![authorizeflow01](https://user-images.githubusercontent.com/13768212/32610823-86e89e60-c529-11e7-94f9-40f584715936.png)
>Authorize

![authorizeflow02](https://user-images.githubusercontent.com/13768212/32610855-a01a7ea8-c529-11e7-8b0f-81759e8e8c49.png)
>signin-AzureAd

![authorizeflow03](https://user-images.githubusercontent.com/13768212/32610858-a2696b10-c529-11e7-97ac-f5d29e4d5052.png)
>ExternalLoginCallback

![authorizeflow04](https://user-images.githubusercontent.com/13768212/32610866-a501e9a6-c529-11e7-8121-c6a45c883a47.png)

**Comment by HaoK** [comment]
Okay well you are pretty close then if you are getting the external cookie, then we just need to figure out why its not getting returned...

So look at https://github.com/aspnet/Identity/blob/rel/2.0.0/src/Microsoft.AspNetCore.Identity/SignInManager.cs#L555

Can you try explicitly calling await Context.AuthenticateAsync(IdentityConstants.ExternalScheme);  since you have the cookie, that should return you a result you can debug into.  It should contain the claims principal from AzureAd as well.

**Comment by AfkAtm** [comment]
This is where it is failing: 
https://github.com/aspnet/Identity/blob/rel/2.0.0/src/Microsoft.AspNetCore.Identity/SignInManager.cs#L577
It would appear that none of the claims are getting populated. So when it looks for the name identifier claim it doesn't find it and returns null. 
![authorizeflow05](https://user-images.githubusercontent.com/13768212/32662437-291a7070-c5f0-11e7-9cb7-60784489672c.png)

Where are these claims populated from the token?

**Comment by AfkAtm** [comment]
Got it working.  My issues were multiple.  
In my initial example I had the resource being set to my application ID, I should have be passing in the graph URL.  
My scope was fine but then I hadn't set up a mapping for the ClaimTypes.NameIdentifier claim.  
Then finally I had to update the User Information Endpoint to point to the same as the Resource Graph uri that I provided.  

Now it works like a charm.  Closing the issue.
